File: android/src/main/java/com/securitysuite/crypto/signatures/EcdsaSigner.java
package com.securitysuite.crypto.signatures;

import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.ECGenParameterSpec;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;

/**
 * ECDSA P-256 digital signatures (SHA-256 digest).
 *
 * Key encoding: DER (PKCS#8 for private, X.509 SubjectPublicKeyInfo for public).
 * Signature: DER-encoded ASN.1 SEQUENCE, matching iOS P256.Signing.ECDSASignature.derRepresentation.
 */
public final class EcdsaSigner {
  private static final String ALGORITHM = "SHA256withECDSA";
  private static final String EC_CURVE = "secp256r1";

  private EcdsaSigner() {}

  public static KeyPair generateKeyPair() throws Exception {
    KeyPairGenerator generator = KeyPairGenerator.getInstance("EC");
    generator.initialize(new ECGenParameterSpec(EC_CURVE));
    return generator.generateKeyPair();
  }

  /** Signs message bytes using SHA-256 / ECDSA P-256 and returns a DER-encoded signature. */
  public static byte[] sign(byte[] message, byte[] privateKeyDer) throws Exception {
    KeyFactory keyFactory = KeyFactory.getInstance("EC");
    PrivateKey privateKey = keyFactory.generatePrivate(new PKCS8EncodedKeySpec(privateKeyDer));
    Signature signer = Signature.getInstance(ALGORITHM);
    signer.initSign(privateKey);
    signer.update(message);
    return signer.sign();
  }

  public static boolean verify(byte[] message, byte[] signatureDer, byte[] publicKeyDer) throws Exception {
    KeyFactory keyFactory = KeyFactory.getInstance("EC");
    PublicKey publicKey = keyFactory.generatePublic(new X509EncodedKeySpec(publicKeyDer));
    Signature verifier = Signature.getInstance(ALGORITHM);
    verifier.initVerify(publicKey);
    verifier.update(message);
    return verifier.verify(signatureDer);
  }

  public static byte[] exportPublicKeyDer(KeyPair keyPair) {
    return keyPair.getPublic().getEncoded();
  }

  public static byte[] exportPrivateKeyDer(KeyPair keyPair) {
    return keyPair.getPrivate().getEncoded();
  }
}

File: ios/Crypto/signatures/EcdsaSigner.swift
import Foundation
import CryptoKit

/**
 * ECDSA P-256 digital signatures (SHA-256 / P-256).
 *
 * Key encoding: DER (PKCS#8-compatible derRepresentation for private, X.509 for public).
 * Signature: DER-encoded ASN.1 SEQUENCE — matches Android SHA256withECDSA output format.
 */
@available(iOS 13.0, *)
struct EcdsaKeyPair {
  let privateKeyDer: Data
  let publicKeyDer: Data
}

@available(iOS 13.0, *)
enum EcdsaSigner {
  static func generateKeyPair() -> EcdsaKeyPair {
    let key = P256.Signing.PrivateKey()
    return EcdsaKeyPair(
      privateKeyDer: key.derRepresentation,
      publicKeyDer: key.publicKey.derRepresentation
    )
  }

  /** Signs message bytes using SHA-256 over P-256 and returns the DER-encoded signature. */
  static func sign(message: Data, privateKeyDer: Data) throws -> Data {
    let privateKey = try P256.Signing.PrivateKey(derRepresentation: privateKeyDer)
    return try privateKey.signature(for: message).derRepresentation
  }

  static func verify(message: Data, signatureDer: Data, publicKeyDer: Data) throws -> Bool {
    let publicKey = try P256.Signing.PublicKey(derRepresentation: publicKeyDer)
    let signature = try P256.Signing.ECDSASignature(derRepresentation: signatureDer)
    return publicKey.isValidSignature(signature, for: message)
  }
}
